File: SexyBiscuit.Engine/Gameplay/GameMode.cs
using Microsoft.Xna.Framework;
using SexyBiscuit.Engine.Core;

namespace SexyBiscuit.Engine.Gameplay;

/// <summary>
/// Marks an actor as a valid spawn location. The game mode picks between all
/// enabled start points when spawning a player.
/// </summary>
public sealed class PlayerStart : Component
{
    /// <summary>Every enabled player start in the scene, in registration order.</summary>
    public static readonly List<PlayerStart> All = new();

    /// <summary>Only starts matching a controller's team are eligible; -1 means any team.</summary>
    public int TeamId { get; set; } = -1;

    /// <summary>Spawn position, taken from the actor's <see cref="Transform3D"/>.</summary>
    public Vector3 Position => Actor.GetComponent<Transform3D>()?.Position ?? Vector3.Zero;

    /// <summary>Spawn yaw in degrees, taken from the actor's <see cref="Transform3D"/>.</summary>
    public float Yaw => Actor.GetComponent<Transform3D>()?.EulerAngles.Y ?? 0f;

    public override void Awake()   => All.Add(this);
    public override void OnDestroy() => All.Remove(this);
}

/// <summary>
/// The rules of a match: who spawns, where, what happens on death, and when the match ends.
/// Modelled on Unreal's <c>AGameModeBase</c>.
/// </summary>
/// <remarks>
/// In a networked session only the server runs a game mode — clients see its consequences
/// through <see cref="GameState"/>. Set <see cref="PawnFactory"/> and
/// <see cref="PlayerControllerFactory"/> to control what gets spawned, or override
/// <see cref="SpawnDefaultPawnFor"/> for full control.
/// </remarks>
/// <example>
/// <code>
/// public class DeathmatchMode : GameMode
/// {
///     public DeathmatchMode()
///     {
///         PawnFactory             = () =&gt; new Character("Fighter");
///         PlayerControllerFactory = () =&gt; new MyPlayerController();
///         ScoreToWin              = 25;
///     }
///
///     public override void OnPlayerDied(PlayerController victim, PlayerController? killer)
///     {
///         killer?.PlayerState?.AddScore(1);
///         RestartPlayer(victim);
///     }
/// }
/// </code>
/// </example>
public class GameMode : Actor
{
    /// <summary>The game mode running in the active scene, or null if none was spawned.</summary>
    public static GameMode? Current { get; private set; }

    /// <summary>Creates the pawn a joining player possesses. Defaults to a plain <see cref="Character"/>.</summary>
    public Func<Pawn> PawnFactory { get; set; } = () => new Character();

    /// <summary>Creates the controller for a joining player.</summary>
    public Func<PlayerController> PlayerControllerFactory { get; set; } = () => new PlayerController();

    /// <summary>Creates the per-player state object.</summary>
    public Func<PlayerState> PlayerStateFactory { get; set; } = () => new PlayerState();

    /// <summary>Creates the match-wide state object.</summary>
    public Func<GameState> GameStateFactory { get; set; } = () => new GameState();

    /// <summary>Layer new gameplay actors are added to.</summary>
    public string SpawnLayer { get; set; } = "default";

    /// <summary>Seconds between a player dying and respawning through <see cref="RestartPlayer"/>.</summary>
    public float RespawnDelay { get; set; } = 3f;

    /// <summary>Score at which <see cref="EndMatch"/> fires automatically. Zero disables it.</summary>
    public int ScoreToWin { get; set; }

    /// <summary>Match length in seconds. Zero means no time limit.</summary>
    public float TimeLimit { get; set; }

    /// <summary>When true the mode spawns one local player as soon as the match starts.</summary>
    public bool AutoStartLocalPlayer { get; set; } = true;

    /// <summary>The match-wide state this mode drives.</summary>
    public GameState GameState { get; private set; } = null!;

    /// <summary>Controllers currently in the match.</summary>
    public List<PlayerController> Controllers { get; } = new();

    /// <summary>Raised when the match begins.</summary>
    public SBEvent MatchStarted { get; } = new();

    /// <summary>Raised when the match ends, carrying the winner or null for a draw.</summary>
    public SBEvent<PlayerState?> MatchEnded { get; } = new();

    private int _nextPlayerId;

    public GameMode() : base("GameMode") { }

    // -------------------------------------------------------------------------
    // Lifecycle
    // -------------------------------------------------------------------------

    protected override void OnStart()
    {
        Current = this;

        // In an editor's edit-time scene nothing is played: no game state, no players. The
        // match starts when Play loads the scene afresh with PlayMode.IsActive set.
        if (!PlayMode.IsActive) return;

        GameState = GameStateFactory();
        Scene?.AddActor(GameState, SpawnLayer);

        OnPreStart();
        StartMatch();
    }

    /// <summary>Called after the game state exists but before the match starts.</summary>
    protected virtual void OnPreStart() { }

    /// <summary>Moves the match into <see cref="MatchState.InProgress"/> and spawns the local player.</summary>
    public void StartMatch()
    {
        if (GameState.MatchState == MatchState.InProgress) return;

        SetMatchState(MatchState.InProgress);
        GameState.ElapsedTime = 0f;

        if (AutoStartLocalPlayer) SpawnPlayer(0);

        OnMatchStart();
        MatchStarted.Broadcast();
    }

    /// <summary>Called once when the match begins.</summary>
    protected virtual void OnMatchStart() { }

    /// <summary>Ends the match and broadcasts the winner.</summary>
    public void EndMatch(PlayerState? winner = null)
    {
        if (GameState.MatchState == MatchState.Ended) return;

        SetMatchState(MatchState.Ended);
        OnMatchEnd(winner);
        MatchEnded.Broadcast(winner);
    }

    /// <summary>Called once when the match ends.</summary>
    protected virtual void OnMatchEnd(PlayerState? winner) { }

    private void SetMatchState(MatchState state)
    {
        GameState.MatchState = state;
        GameState.MatchStateChanged.Broadcast(state);
    }

    // -------------------------------------------------------------------------
    // Player lifecycle
    // -------------------------------------------------------------------------

    /// <summary>
    /// Creates a controller, player state and pawn for a local player and possesses the pawn.
    /// </summary>
    /// <param name="playerIndex">Local player slot — 0 for the first player, 1 for split-screen player two.</param>
    public virtual PlayerController SpawnPlayer(int playerIndex = 0)
    {
        var controller = PlayerControllerFactory();
        controller.PlayerIndex = playerIndex;

        var state = PlayerStateFactory();
        state.PlayerId   = _nextPlayerId++;
        state.PlayerName = $"Player {state.PlayerId + 1}";
        controller.PlayerState = state;

        Scene?.AddActor(state,      SpawnLayer);
        Scene?.AddActor(controller, SpawnLayer);

        Controllers.Add(controller);
        GameState.AddPlayer(state);

        var pawn = SpawnDefaultPawnFor(controller);
        if (pawn != null) controller.Possess(pawn);

        OnPlayerJoined(controller);
        return controller;
    }

    /// <summary>
    /// Creates the pawn for <paramref name="controller"/> and places it at a chosen
    /// <see cref="PlayerStart"/>. Override to spawn different pawns per team or class.
    /// </summary>
    public virtual Pawn? SpawnDefaultPawnFor(PlayerController controller)
    {
        var pawn = PawnFactory();
        var t3d  = pawn.GetComponent<Transform3D>() ?? pawn.AddComponent<Transform3D>();

        var start = ChoosePlayerStart(controller);
        if (start != null)
        {
            t3d.Position     = start.Position;
            t3d.EulerAngles  = new Vector3(0f, start.Yaw, 0f);
            pawn.ControlRotation = new Vector3(0f, start.Yaw, 0f);
        }

        Scene?.AddActor(pawn, SpawnLayer);
        return pawn;
    }

    /// <summary>
    /// Picks a spawn point for <paramref name="controller"/>. The default picks the
    /// team-matching start furthest from any other player, falling back to the first start.
    /// </summary>
    protected virtual PlayerStart? ChoosePlayerStart(PlayerController controller)
    {
        int team = controller.PlayerState?.TeamId ?? -1;

        var eligible = PlayerStart.All
            .Where(s => s.Actor.IsActive && (s.TeamId < 0 || team < 0 || s.TeamId == team))
            .ToList();

        if (eligible.Count == 0) return null;
        if (eligible.Count == 1) return eligible[0];

        var occupied = Controllers
            .Select(c => c.ControlledPawn?.GetComponent<Transform3D>()?.Position)
            .Where(p => p.HasValue)
            .Select(p => p!.Value)
            .ToList();

        if (occupied.Count == 0) return eligible[0];

        return eligible
            .OrderByDescending(s => occupied.Min(o => Vector3.DistanceSquared(o, s.Position)))
            .First();
    }

    /// <summary>
    /// Destroys the controller's current pawn and spawns a fresh one after
    /// <see cref="RespawnDelay"/> seconds.
    /// </summary>
    public virtual void RestartPlayer(PlayerController controller)
    {
        controller.ControlledPawn?.Destroy();
        controller.UnPossess();

        TimerManager.Instance.SetTimer(RespawnDelay, () =>
        {
            if (GameState.MatchState != MatchState.InProgress) return;
            var pawn = SpawnDefaultPawnFor(controller);
            if (pawn != null) controller.Possess(pawn);
        });
    }

    /// <summary>Removes a player from the match and destroys their actors.</summary>
    public virtual void RemovePlayer(PlayerController controller)
    {
        controller.ControlledPawn?.Destroy();
        controller.UnPossess();
        Controllers.Remove(controller);

        if (controller.PlayerState != null)
        {
            GameState.RemovePlayer(controller.PlayerState);
            controller.PlayerState.Destroy();
        }

        controller.Destroy();
    }

    /// <summary>Called after a player joins and has been given a pawn.</summary>
    protected virtual void OnPlayerJoined(PlayerController controller) { }

    /// <summary>
    /// Called by gameplay code when a player dies. The default awards the killer a point
    /// and restarts the victim.
    /// </summary>
    public virtual void OnPlayerDied(PlayerController victim, PlayerController? killer)
    {
        killer?.PlayerState?.AddScore(1);
        RestartPlayer(victim);
    }

    // -------------------------------------------------------------------------
    // Win conditions
    // -------------------------------------------------------------------------

    protected override void Update(float dt)
    {
        if (GameState.MatchState != MatchState.InProgress) return;

        if (TimeLimit > 0f && GameState.ElapsedTime >= TimeLimit)
        {
            EndMatch(GameState.GetScoreboard().FirstOrDefault());
            return;
        }

        if (ScoreToWin > 0)
        {
            var leader = GameState.GetScoreboard().FirstOrDefault();
            if (leader != null && leader.Score >= ScoreToWin) EndMatch(leader);
        }
    }

    protected override void OnDestroy()
    {
        if (ReferenceEquals(Current, this)) Current = null;
    }
}

File: html5/src/gameplay/GameMode.js
// -----------------------------------------------------------------------------
// GameMode — the rules: who spawns, where, and when the match ends.
//
// It waits for PlayMode.isActive before doing anything, so opening a scene in
// the editor does not populate it with controllers and pawns that then get
// serialised into the file.
// -----------------------------------------------------------------------------

import { Actor } from '../core/Actor.js';
import { registerActor } from '../core/TypeRegistry.js';
import { PropertyType as P } from '../core/PropertyTypes.js';
import { SBEvent } from '../core/SBEvent.js';
import { PlayMode } from '../core/PlayMode.js';
import { Vector3, Color } from '../math/index.js';
import { Transform3D } from '../core/Transform3D.js';
import { MeshRenderer } from '../rendering/MeshRenderer.js';
import { MeshPrimitive } from '../rendering/PrimitiveMesh.js';
import { Camera3D } from '../rendering/Camera3D.js';
import { Character } from './Character.js';
import { PlayerController } from './PlayerController.js';
import { PlayerState } from './PlayerState.js';
import { GameState, MatchState } from './GameState.js';
import { PlayerStart } from './PlayerStart.js';
import { CharacterController3D } from '../physics/CharacterController3D.js';

/** The match rules. One per scene. */
export class GameMode extends Actor {
    static schema = {
        spawnLayer:          { type: P.String, default: 'default' },
        respawnDelay:        { type: P.Number, default: 3 },
        scoreToWin:          { type: P.Int, default: 0 },
        timeLimit:           { type: P.Number, default: 0 },
        autoStartLocalPlayer:{ type: P.Bool, default: true },
    };

    /** The game mode running in the active scene. */
    static current = null;

    constructor(name = 'GameMode') {
        super(name);

        /** Factories, so a subclass can swap in its own classes without overriding spawn. */
        this.pawnFactory = () => new Character();
        this.playerControllerFactory = () => new PlayerController();
        this.playerStateFactory = () => new PlayerState();
        this.gameStateFactory = () => new GameState();

        this.spawnLayer = 'default';
        this.respawnDelay = 3;
        /** Zero means no score limit. */
        this.scoreToWin = 0;
        /** Zero means no time limit, in seconds. */
        this.timeLimit = 0;
        this.autoStartLocalPlayer = true;

        /** @type {?GameState} */
        this.gameState = null;

        /** @type {PlayerController[]} */
        this.controllers = [];

        this.matchStarted = new SBEvent();
        this.matchEnded = new SBEvent();

        this._started = false;
    }

    onStart() {
        // Edit mode: the scene is being authored, not played. Spawning here would
        // put controllers and pawns into the file the editor is about to save.
        if (!PlayMode.isActive) return;
        this._beginPlay();
    }

    update(dt) {
        if (!PlayMode.isActive) return;
        if (!this._started) { this._beginPlay(); return; }

        if (this.gameState?.matchState !== MatchState.InProgress) return;

        if (this.timeLimit > 0 && this.gameState.elapsedTime >= this.timeLimit) {
            this.endMatch(this.gameState.getScoreboard()[0] ?? null);
        }
    }

    _beginPlay() {
        this._started = true;
        GameMode.current = this;

        this.gameState = this.gameStateFactory();
        this.scene?.addActor(this.gameState, this.spawnLayer);

        this.onPreStart();

        if (this.autoStartLocalPlayer) this.spawnPlayer(0);

        this.startMatch();
    }

    /** Moves the match into progress. */
    startMatch() {
        if (!this.gameState) return;
        this.gameState.matchState = MatchState.InProgress;
        this.onMatchStart();
        this.matchStarted.broadcast();
    }

    /** Ends the match, optionally naming a winner. */
    endMatch(winner = null) {
        if (!this.gameState || this.gameState.matchState === MatchState.Ended) return;
        this.gameState.matchState = MatchState.Ended;
        this.onMatchEnd(winner);
        this.matchEnded.broadcast(winner);
    }

    /** Creates a controller, its state and its pawn, and possesses. */
    spawnPlayer(playerIndex = 0) {
        const controller = this.playerControllerFactory();
        controller.playerIndex = playerIndex;
        controller.name = `PlayerController ${playerIndex}`;

        const playerState = this.playerStateFactory();
        playerState.playerId = playerIndex;
        playerState.playerName = `Player ${playerIndex + 1}`;
        controller.playerState = playerState;

        this.scene?.addActor(playerState, this.spawnLayer);
        this.scene?.addActor(controller, this.spawnLayer);
        this.gameState?.addPlayer(playerState);
        this.controllers.push(controller);

        // The actors are queued; flushing now means the pawn can be possessed in
        // this call rather than a frame later, when a caller expects a live pawn.
        this.scene?.flushPendingActors();

        const pawn = this.spawnDefaultPawnFor(controller);
        if (pawn) controller.possess(pawn);

        this.onPlayerJoined(controller);
        return controller;
    }

    /** Creates the pawn for a controller and places it at a spawn point. */
    spawnDefaultPawnFor(controller) {
        const pawn = this.pawnFactory();
        if (!pawn) return null;

        pawn.name = `Pawn ${controller.playerIndex}`;
        this.scene?.addActor(pawn, this.spawnLayer);
        this.scene?.flushPendingActors();

        const start = this.choosePlayerStart(controller);
        const t = pawn.transform3D;
        if (t && start) {
            // A Player Start marks where the pawn's *feet* go. A capsule's
            // transform sits at its middle, so dropping it straight onto the
            // marker buries it up to the waist — and a character that starts
            // inside the floor never finds ground to stand on, because the floor
            // is above its feet. It falls forever, which is what pressing Play on
            // the default scene did.
            const footOffset = pawn.getComponent(CharacterController3D)?.footOffset ?? 0;
            const feet = start.position;

            t.position = new Vector3(feet.x, feet.y + footOffset, feet.z);
            pawn.controlRotation = new Vector3(0, start.yaw, 0);
            t.eulerAngles = new Vector3(0, start.yaw, 0);
        }

        this.equipDefaultPawn(pawn);
        return pawn;
    }

    /**
     * Gives a pawn a body and a view, if it arrived with neither.
     *
     * The default pawn is a bare `Character`: no renderer, no camera. It walks,
     * falls and collides correctly, and none of that is visible — pressing Play
     * on a fresh scene looked exactly like not pressing it, which is a reasonable
     * reason to conclude that Play is broken.
     *
     * Both additions are conditional, so a game that builds its own pawn through
     * `pawnFactory` keeps whatever rig it supplied. Both hang off the pawn as
     * child actors rather than sitting on it: the body has to be scaled to match
     * the capsule the controller moves, and scaling the pawn itself would drag
     * the camera in with it.
     */
    equipDefaultPawn(pawn) {
        const root = pawn.transform3D;
        if (!root) return;

        const movement = pawn.getComponent(CharacterController3D);

        if (!pawn.getComponent(MeshRenderer) && !this._findInChildren(root, MeshRenderer)) {
            const body = new Actor(`${pawn.name} Body`);
            const bodyTransform = body.addComponent(Transform3D);

            const renderer = body.addComponent(MeshRenderer);
            renderer.setPrimitive(MeshPrimitive.Capsule);
            renderer.albedoColor = Color.from('#D98C4A');
            renderer.roughness = 0.55;

            this._addChild(body, bodyTransform, root);

            // The capsule primitive is a metre across and two tall; scale it to
            // whatever the controller is actually sweeping.
            if (movement) {
                bodyTransform.localScale = new Vector3(
                    movement.radius * 2, movement.height / 2, movement.radius * 2);
            }
        }

        if (PlayerController.findPawnCamera(pawn)) return;

        // Over the shoulder, behind the pawn. Behind is +Z, because the engine's
        // forward is -Z; with no yaw of its own the camera inherits the pawn's
        // facing and so looks along -Z, back across the pawn and out in front of
        // it. Yawing it 180 here would point it at the horizon behind the player.
        const view = new Actor(`${pawn.name} Camera`);
        const viewTransform = view.addComponent(Transform3D);
        view.addComponent(Camera3D);

        this._addChild(view, viewTransform, root);

        const eye = movement ? movement.footOffset + 0.7 : 1.6;
        viewTransform.localPosition = new Vector3(0, eye, 4.5);
        viewTransform.localEulerAngles = new Vector3(-10, 0, 0);
    }

    /** Adds an actor to the scene and parents it, keeping its local placement. */
    _addChild(actor, transform, parent) {
        this.scene?.addActor(actor, this.spawnLayer);
        // Flush before parenting: an actor still queued has not run `start`, and
        // its components have nothing to attach to yet.
        this.scene?.flushPendingActors();
        transform.setParent(parent, false);
    }

    /** The first component of a type anywhere beneath a transform. */
    _findInChildren(root, type) {
        const stack = [...root.children];
        while (stack.length > 0) {
            const transform = stack.pop();
            const found = transform.actor?.getComponent(type);
            if (found) return found;
            stack.push(...transform.children);
        }
        return null;
    }

    /** Picks a spawn point for a controller. */
    choosePlayerStart(controller) {
        const starts = PlayerStart.all.filter((s) => s.actor?.scene === this.scene);
        if (starts.length === 0) return null;

        const team = controller.playerState?.teamId ?? -1;
        const matching = starts.filter((s) => s.teamId === -1 || s.teamId === team);
        const pool = matching.length > 0 ? matching : starts;

        // Round-robin rather than random: two players joining together should not
        // land on top of each other.
        return pool[controller.playerIndex % pool.length];
    }

    /** Destroys and re-spawns a controller's pawn. */
    restartPlayer(controller) {
        controller.controlledPawn?.destroy();
        controller.unPossess();

        const pawn = this.spawnDefaultPawnFor(controller);
        if (pawn) controller.possess(pawn);
    }

    /** Removes a player from the match entirely. */
    removePlayer(controller) {
        const i = this.controllers.indexOf(controller);
        if (i >= 0) this.controllers.splice(i, 1);

        if (controller.playerState) this.gameState?.removePlayer(controller.playerState);
        controller.controlledPawn?.destroy();
        controller.playerState?.destroy();
        controller.destroy();
    }

    /** Scores the kill and schedules a respawn. */
    onPlayerDied(victim, killer = null) {
        if (killer && killer !== victim) {
            killer.playerState?.addScore(1);
            if (this.scoreToWin > 0 && (killer.playerState?.score ?? 0) >= this.scoreToWin) {
                this.endMatch(killer.playerState);
                return;
            }
        }

        victim.controlledPawn?.destroy();
        victim.unPossess();

        const timers = this.scene?.engine?.timers;
        if (timers) {
            timers.setTimer(this.respawnDelay, () => {
                if (!victim.isDestroyed) this.restartPlayer(victim);
            });
        } else {
            this.restartPlayer(victim);
        }
    }

    onDestroy() {
        if (GameMode.current === this) GameMode.current = null;
    }

    // ---- Hooks for subclasses ------------------------------------------------

    /** Runs before any player is spawned. */
    onPreStart() {}
    onMatchStart() {}
    onMatchEnd(winner) {}
    onPlayerJoined(controller) {}
}
registerActor(GameMode, { category: 'Gameplay', summary: 'The match rules.' });
